### ISSUE TITLE
Feature to discuss: create xfce4-panel.xml by nix

### ISSUE THREAD

**Issue by volth** [NixOS/nixpkgs#35249]
It might be useful to use Nix to define XFCE panel: 
 * which plugins are to be displayed
 * which programs launcher buttons would launch
 * which languages the language switched to work with
 * which timezones the clocks will display time in
 * ....

Configuring XFCE panel manually on each new laptop / workstation / usb-flash drive / ... is annoying.
 
@yegortimoshenko @vcunat

**Comment by volth** [comment]
Currently I have it kind of automated by 
```
services.xserver.desktopManager.xfce.extraSessionCommands = ''
  rm -rf $HOME/.config/xfce4 || true
  cp -dpR ${./config-xfce4} $HOME/.config/xfce4
  chmod -R +w $HOME/.config/xfce4
'';
```
but as the number of machines grows and their panel configs are suppose to be slightly different (due to absent of some software or different language/time preferences) supporting per-machine directories ```./config-xfce4-${config.networking.hostName}``` full of xml files seems to be an antipattern.

**Comment by yegortimoshenko** [comment]
That's definitely very annoying, both setting it up on each machine and also syncing changes.

It can be argued that NixUP or home-manager is a better way to solve this problem, but the former is vaporware and the latter would substantially diminish reach. It's also not clear how fully decoupled declarative userspace configuration manager would resolve compatibility between Xfce versions (config format might not be compatible).

I think that this should follow `time.timeZone` design: i.e. imperative configuration by default, and optional declarative configuration (done via NixOS options) that disables imperative UI (i.e. corresponding Settings app, or in case with panels, panel preferences).

This approach would apply well to most Xfce settings, including panel plugins, i.e. launcher buttons, language switcher, clock timezone...

Implementing that is going to require some patches, which I would be happy to work on.

**Comment by volth** [comment]
Per-user config, yes, this might be useful too (although nowadays it is more typical one user owning many devices than one device shared by many users).

Even without NixUP/home-manager, ```nixpkgs``` has some per-user settings (such as ```initialPassword```) so per-userness of panel config can be solved in as similar way.

**Comment by vcunat** [comment]
Yes, I also see it in the NixUP/home-manager family.  My approach in the past few years is using VCS (git) over a subset of dotfiles (`~/.**`).

**Comment by volth** [comment]
It is not about preserving the dotfiles, it is about getting rid of them.
And make their (former) content configurable via NixOS options and depended on the other.

Something like (pseudocode, not a proposed Nix):
```
xfce.panel = [
    ...
    (xfce.panel.plugin.clock { timezone = "UTC" })
 ++ (optional (config.system.timeZone != "UTC")
        xfce.panel.plugin.clock { timezone = config.system.timeZone })
 ++ [ (xfce.panel.plugin.langswitcher { languages = [ "EN" "DE" "RU" ]; }) ]
];
```

**Comment by vcunat** [comment]
The hard part there is that you want some (bidirectional) mapping from nix to various kinds of configuration files, so someone would need to write them for all kinds of software.

**Comment by volth** [comment]
@vcunat 
> you want some (bidirectional) mapping from..

Why "bidirectional" ?
Only from Nix to dotfiles (and perhaps to patches if something is not configurable via dotfiles), why back?

> version control on dotfiles is fairly easy and universal.

That is what I am using now, keeping dotfiles of each machine in git along with .nix-files describing configuration.

**Comment by vcunat** [comment]
If you have accumulated some dotfiles, you might like to convert them to nix when you switch to nix-managing them.  I parenthesized it to mean lower importance of the other direction.

**Comment by vcunat** [comment]
BTW, I have one repo with dotfiles for all machines, with common stuff committed to master (almost everything) and a few specific changes added to each machine's branch.
